xcm-handler/src/lib.rs

```diff
@@ -0,0 +1,199 @@
+		}
+
+		#[weight = 1_000]
+		fn sudo_send_hrmp_xcm(origin, recipient: ParaId, message: VersionedXcm) {
+			ensure_root(origin)?;
+			let data = message.encode();
+			let outbound_message = OutboundHrmpMessage {
+				recipient,
+				data,
+			};
+			T::HrmpMessageSender::send_hrmp_message(outbound_message).map_err(|_| Error::<T>::FailedToSend)?;
+		}
+	}
+}
+
+impl<T: Config> DownwardMessageHandler for Module<T> {
+	fn handle_downward_message(msg: InboundDownwardMessage) {
+		let hash = msg.using_encoded(T::Hashing::hash);
+		frame_support::debug::print!("Processing Downward XCM: {:?}", &hash);
+		match VersionedXcm::decode(&mut &msg.msg[..]).map(Xcm::try_from) {
+			Ok(Ok(xcm)) => {
+				match T::XcmExecutor::execute_xcm(Junction::Parent.into(), xcm) {
+					Ok(..) => RawEvent::Success(hash),
+					Err(e) => RawEvent::Fail(hash, e),
+				};
+			}
+			Ok(Err(..)) => Self::deposit_event(RawEvent::BadVersion(hash)),
+			Err(..) => Self::deposit_event(RawEvent::BadFormat(hash)),
+		}
+	}
+}
+
+impl<T: Config> HrmpMessageHandler for Module<T> {
+	fn handle_hrmp_message(sender: ParaId, msg: InboundHrmpMessage) {
+		let hash = msg.using_encoded(T::Hashing::hash);
+		frame_support::debug::print!("Processing HRMP XCM: {:?}", &hash);
+		match VersionedXcm::decode(&mut &msg.data[..]).map(Xcm::try_from) {
+			Ok(Ok(xcm)) => {
+				match T::XcmExecutor::execute_xcm(Junction::Parachain { id: sender.into() }.into(), xcm) {
+					Ok(..) => RawEvent::Success(hash),
+					Err(e) => RawEvent::Fail(hash, e),
+				};
+			}
+			Ok(Err(..)) => Self::deposit_event(RawEvent::BadVersion(hash)),
+			Err(..) => Self::deposit_event(RawEvent::BadFormat(hash)),
+		}
+	}
+}
+
+impl<T: Config> SendXcm for Module<T> {
+	fn send_xcm(dest: MultiLocation, msg: Xcm) -> Result<(), XcmError> {
+		let msg: VersionedXcm = msg.into();
+		match dest.first() {
+			// A message for us. Execute directly.
+			None => {
+				let msg = msg.try_into().map_err(|_| XcmError::UnhandledXcmVersion)?;
+				let res = T::XcmExecutor::execute_xcm(MultiLocation::Null, msg);
+				res
+			}
+			// An upward message for the relay chain.
+			Some(Junction::Parent) if dest.len() == 1 => {
+				let data = msg.encode();
+				let hash = T::Hashing::hash(&data);
+
+				T::UpwardMessageSender::send_upward_message(data).map_err(|_| XcmError::Undefined)?;
+				Self::deposit_event(RawEvent::UpwardMessageSent(hash));
+
+				Ok(())
+			}
+			// An HRMP message for a sibling parachain.
+			Some(Junction::Parachain { id }) => {
+				let data = msg.encode();
+				let hash = T::Hashing::hash(&data);
+				let message = OutboundHrmpMessage {
+					recipient: (*id).into(),
+					data,
+				};
+				// TODO: Better error here
+				T::HrmpMessageSender::send_hrmp_message(message).map_err(|_| XcmError::Undefined)?;
+				Self::deposit_event(RawEvent::HrmpMessageSent(hash));
+				Ok(())
+			}
+			_ => {
+				/* TODO: Handle other cases, like downward message */
```

Comment:
How is a downward message possible? We are already at the "bottom" isn't it?

Reply:
This might be a note for like future nested relay chain support.